docs/rfcs/001-updateable-bundled-packages.md

```diff
@@ -0,0 +1,100 @@
+
+### Implementation Details
+
+Because the necessary infrastructure is already in place to enable updates to bundled packages using `apm`, the only work required is to provide a way for packages to opt in to this behavior and for `apm` to include those packages in its update checks if they haven't already been installed in the user's packages folder.
+
+Any bundled Atom package will be able to opt in to updates by adding `"updateable": true` to its `package.json` file.  This will cause `apm` to consider it as part of the list of packages it checks for updates.  If a community (non-bundled) package sets this field to `true` or `false` it will be ignored as it's only relevant to bundled packages.
+
+`apm` will be updated to include the list of bundled packages with `"updateable": true` set in their `package.json` so that the user will be notified of new package versions that support the engine version of their current Atom build.
+
+### User Experience Examples
+
+1. The user downloads Atom 1.28.0 from atom.io which includes GitHub package version 0.15.0.  After Atom 1.28.0 was released, a hotfix release was shipped for the GitHub package as 0.15.1.  When the user installs and starts Atom, they are prompted to install the update to the GitHub package.
+
+2. The user downloads and installs Atom 1.28.0 from atom.io which includes GitHub package version 0.15.0.  Two weeks later, GitHub package 0.16.0 is released with a few new features.  The user is prompted to update to the new version and gets the new features even though Atom 1.29.0 hasn't been released yet.
+
+3. In the future, a user has an old install of Atom 1.28.0 and waits a long time between installing Atom updates.  The GitHub package releases version 0.25.0 but the user is not prompted to install it because the GitHub package has set `engines` in `package.json` to restrict to Atom 1.32.0 and above.
+
+### Rules for Updateable Bundled Packages
+
+Any package that opts into this behavior must follow one rule: **its `engines` field must be regularly updated to reflect the necessary Atom version for the Atom, Electron, and Node.js APIs used in the package**.  This field defines the range of Atom versions in which the package is expected to work.  The field should always be set to the lowest possible Atom version that the package supports.
+
+If a package wants to use API features of a newer version of Atom while still supporting older Atom versions, it must do so in a way that is aware of the user's version and adjust itself accordingly.
+
+## Drawbacks
+
+> Why should we *not* do this?
+
+The primary drawback of this approach is that updateable bundled packages might exhibit problems on older Atom versions due to missing or changed APIs in Atom, Electron, or Node.js.  The solution for these packages is to keep their `engines` field updated appropriately, but there's still a chance that some updates will slip through without the necessary engine version changes.
```

Comment:
The single biggest concern for me is how out-of-band updates would impact v8 snapshotting of the github package in particular. Are we going to try to share *some* snapshotted code but not all, switch to a completely different strategy of computing startup snapshots where we ship `mksnapshot` and actually compute them on the user's machine, or just lose the benefit of startup snapshots if the user upgrades their bundled package.

None of those sound very appealing to me. Are there alternatives I haven't thought of?

Reply:
It would take a lot of work in APM and Atom to handle it (and probably an even newer version of Electron? This shipped in Chrome 65), but [Lazy deserialization](https://v8project.blogspot.com/2018/02/lazy-deserialization.html) at least _sounds_ like it might be a good fit for that issue if we can somehow ship snapshots for each package.

Reply:
Lazy deserialization seems like it will benefit us, but I think it's orthogonal to the concern about how to update Atom's startup snapshot when a bundled packages is updated, right?

Reply:
Mostly, there's no way to update the startup snapshot without bundling `mksnapshot` and regenerating it every time a package installs/updates/is removed. Assuming I'm not completely wrong with how that works though, once support for that is in the Electron version we are using I don't think we would _need_ to regenerate the startup snapshot, as each package would provide its own snapshot that would be loaded on the fly.

That's all way in the future enhancements though, I think your concern here is very valid and there isn't currently anyway around losing the benefit of the startup snapshot for packages if they are updated outside of core.

Reply:
Thanks for bringing this up @maxbrunsfeld, I will capture this concern in the RFC because it's likely to be a problem.  Before I finish up the document I'll do some research on how much start up time is impacted by using a non-snapshotted GitHub package then I'll add it here.

If we decide that the cost of not being able to update the snapshot is too high to consider making this change right now, I'm totally on board with postponing or rejecting this RFC and looking for a better overall approach.

Definitely curious to look into @Arcanemagus's suggestion as a future optimization when we're running on more recent Chrome relases.